crates/node/src/storage/rocksdb.rs

```diff
@@ -688,12 +734,272 @@ impl RocksDB {
+                        .and_then(|h| BlockHeight::from_str(h).ok())
+                    else {
+                        continue;
+                    };
+                    // check if we have found the metadata file or snapshot file
+                    // for a given block height
+                    if entry_ext == Some(meta) {
+                        let metadata = std::fs::read_to_string(entry_path)?;
+                        let metadata_bytes = HEXLOWER
+                            .decode(metadata.as_bytes())
+                            .map_err(|e| {
+                                std::io::Error::new(ErrorKind::InvalidData, e)
+                            })?;
+                        let chunks: Vec<Chunk> =
+                            BorshDeserialize::try_from_slice(
+                                &metadata_bytes[..],
+                            )?;
+                        files.entry(height).or_default().1 = Some(chunks);
+                    } else {
+                        files.entry(height).or_default().0 = Some(
+                            base_dir
+                                .join(format!("snapshot_{}", height))
+                                .to_string_lossy()
+                                .into(),
+                        );
+                    }
+                };
+            }
+        }
+        let mut res = Vec::with_capacity(files.len());
+        for (height, (path, chunks)) in files {
+            // only include snapshots which have both a .snap and .meta file.
+            if let Some((path_stem, chunks)) = path.zip(chunks) {
+                res.push(SnapshotMetadata {
+                    height,
+                    path_stem,
+                    chunks,
+                });
+            }
+        }
+        Ok(res)
+    }
+
+    /// Create a path to save a snapshot at a specific block height.
+    pub fn paths(height: BlockHeight, base_dir: PathBuf) -> [PathBuf; 2] {
+        let snap_file = base_dir.join(format!("snapshot_{}.snap", height));
+        let metadata_file = base_dir.join(format!("snapshot_{}.meta", height));
+        [snap_file, metadata_file]
+    }
+
+    /// Load the specified chunk of a snapshot at the given block height
+    pub fn load_chunk(
+        height: BlockHeight,
+        chunk: u64,
+        base_dir: &Path,
+    ) -> std::io::Result<Vec<u8>> {
+        let files = Self::files(base_dir)?;
+        let Some(metadata) = files.into_iter().find(|m| m.height == height)
+        else {
+            return Err(std::io::Error::new(
+                ErrorKind::NotFound,
+                format!(
+                    "Could not find the metadata file for the snapshot at \
+                     height {}",
+                    height,
+                ),
+            ));
+        };
+        let chunk_start = if chunk == 0 {
+            0usize
+        } else {
+            let prev = checked!(usize::try_from(chunk).unwrap() - 1).unwrap();
```

Comment:
we should be careful with these checked arithmetic unwraps. in theory cometbft won't provide bogus values, but they could crash the ledger